pkg/server/plugin/bundlepublisher/awsrolesanywhere/awsrolesanywhere.go

```diff
@@ -0,0 +1,205 @@
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const (
+	pluginName = "aws_rolesanywhere_trustanchor"
+)
+
+type pluginHooks struct {
+	newRolesAnywhereClientFunc func(c aws.Config) (rolesAnywhere, error)
+}
+
+func BuiltIn() catalog.BuiltIn {
+	return builtin(New())
+}
+
+func New() *Plugin {
+	return newPlugin(newRolesAnywhereClient)
+}
+
+// Config holds the configuration of the plugin.
+type Config struct {
+	AccessKeyID     string `hcl:"access_key_id" json:"access_key_id"`
+	SecretAccessKey string `hcl:"secret_access_key" json:"secret_access_key"`
+	Region          string `hcl:"region" json:"region"`
+	TrustAnchorID   string `hcl:"trust_anchor_id" json:"trust_anchor_id"`
+}
+
+// Plugin is the main representation of this bundle publisher plugin.
+type Plugin struct {
+	bundlepublisherv1.UnsafeBundlePublisherServer
+	configv1.UnsafeConfigServer
+
+	config    *Config
+	configMtx sync.RWMutex
+
+	bundle    *types.Bundle
+	bundleMtx sync.RWMutex
+
+	hooks               pluginHooks
+	rolesAnywhereClient rolesAnywhere
+	log                 hclog.Logger
+}
+
+// SetLogger sets a logger in the plugin.
+func (p *Plugin) SetLogger(log hclog.Logger) {
+	p.log = log
+}
+
+// Configure configures the plugin.
+func (p *Plugin) Configure(ctx context.Context, req *configv1.ConfigureRequest) (*configv1.ConfigureResponse, error) {
+	config, err := parseAndValidateConfig(req.HclConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	awsCfg, err := newAWSConfig(ctx, config)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create client configuration: %v", err)
+	}
+	rolesAnywhere, err := p.hooks.newRolesAnywhereClientFunc(awsCfg)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to create client: %v", err)
+	}
+	p.rolesAnywhereClient = rolesAnywhere
+
+	p.setConfig(config)
+	p.setBundle(nil)
+	return &configv1.ConfigureResponse{}, nil
+}
+
+// PublishBundle puts the bundle in the first Roles Anywhere trust anchor
+// found with the configured name. If one doesn't exist, it is created.
+func (p *Plugin) PublishBundle(ctx context.Context, req *bundlepublisherv1.PublishBundleRequest) (*bundlepublisherv1.PublishBundleResponse, error) {
+	config, err := p.getConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	if req.Bundle == nil {
+		return nil, status.Error(codes.InvalidArgument, "missing bundle in request")
+	}
+
+	currentBundle := p.getBundle()
+	if proto.Equal(req.GetBundle(), currentBundle) {
+		// Bundle not changed. No need to publish.
+		return &bundlepublisherv1.PublishBundleResponse{}, nil
+	}
+
+	formatter := bundleformat.NewFormatter(req.GetBundle())
+	bundleBytes, err := formatter.Format(bundleformat.PEM)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "could not format bundle to PEM format")
+	}
+	bundleStr := string(bundleBytes)
+
+	// Update the trust anchor that was found
+	updateTrustAnchorInput := rolesanywhere.UpdateTrustAnchorInput{
```

Comment:
The [maximum length of the SourceData is 8000](https://docs.aws.amazon.com/rolesanywhere/latest/APIReference/API_SourceData.html). Do you think that it would be problematic in your use case?
That would be around 11 certs in the bundle when no upstream authority is used. I think that we should have a validation, and if it's larger than 8000 fail with a descriptive error before trying to update, because the error response from the AWS API will include the SourceData, which can flood the logs.

Reply:
Good catch, I can add that check in. But I think what will usually be more limiting is the fact that the maximum bundle size that can be used for a trust anchor is [two currently](https://docs.aws.amazon.com/rolesanywhere/latest/userguide/quotas.html) (see the "Certificates per trust anchor" limit), and it isn't adjustable either. For example, when I try to update a trust anchor with a bundle of size six, I get the following error: 

```
An error occurred (ValidationException) when calling the UpdateTrustAnchor operation: Unable to update a TrustAnchor with a certificate bundle of size 6.
```
At least the error response doesn't contain the bundle contents, so maybe the concern about flooding the logs isn't as relevant here, but it's still an error nonetheless. I think the intent behind this limit being two is to support CA certificate rotation, without loss of availability. So in general, the service is expecting a single CA certificate to be uploaded as a trust anchor (and two when the CA certificate needs to be rotated). 

Given this, in what scenarios will the trust bundle contain more than a single certificate? And in those cases, would it be okay to have this plugin publish just one of those certificates to the Roles Anywhere trust anchor? I'm guessing this behavior would make the plugin inconsistent with the functionality of the other bundle publisher plugins though - I'm not sure if that's okay. If we do go with this approach and additional certificates are needed, they can be provided in the request to `CreateSession` (the Roles Anywhere credentialing API), in the `X-Amz-X509-Chain` HTTP header (I have to test whether there's a limit here, but it should definitely be greater than one). 

What are your thoughts on this?

Reply:
I think that should work for now, thanks. 

How do you suggest I check that a plugin of a different type has also been configured? Is there any precedence for it? I could potentially "fake" it by adding another configuration field for this plugin, which will allow the user to acknowledge (so a boolean field) that they are also using an `UpstreamAuthority` plugin. It could also be a more general acknowledgement field, so that the user will only configure the plugin if they know for sure they want to use it despite the limitations.

Reply:
Unfortunately, there is no easy way for a plugin to check if other plugin has been configured. Having a setting to explicitly tell that there is an UpstreamAuthority plugin may work, but at the same time it feels a little confusing to me. I would suggest that we start by having a very clear documentation about this plugin being only supported when an UpstreamAuthority plugin is used.

Reply:
Hey @ajay1135 I just wanted to provide a quick summary of what I think is pending:
1) Document that this plugin is only supported when an UpstreamAuthority plugin is used. Probably with a bold note at the very beginning of the .md document.
2) Check that the bundle has 1) no more than 2 certs before is being published and 2) doesn't exceed the maximum length of 8000.
3) Have 2) covered by unit tests.

Reply:
Thanks for summarizing this. Quick question though - do you think it's necessary to explicitly handle the error case when there are more than two certificates in the bundle, before calling the API? In this case, the error message from the API won't flood the logs with the contents of the bundle. Here's what I get when I call the `UpdateTrustAnchor` API through the CLI, with a bundle with three certificates in it: 

```
An error occurred (ValidationException) when calling the UpdateTrustAnchor operation: Unable to update a TrustAnchor with a certificate bundle of size 3.
```

For now, I'm not explicitly handling this error scenario, but if you think I should, I'll go back and update the PR. But I think my new changes do address the rest of your comments.